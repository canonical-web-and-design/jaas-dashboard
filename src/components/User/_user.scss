@import "../../scss/settings";
@import "vanilla-framework/scss/vanilla";

.user {
  @include vf-animation(all, brisk, ease-in-out);

  background-color: $color-navigation-dark;
  bottom: 0;
  color: $color-light;
  max-height: 3rem;
  padding: 0.5rem 0;
  position: absolute;
  width: 100%;

  &.is-expanded {
    max-height: 120px;
  }

  &__header {
    cursor: pointer;
    display: flex;
    padding: 0 1rem;

    [class^="p-icon"] {
      position: relative;
      top: 0.25rem;
    }

    .p-icon--contextual-menu {
      @at-root {
        .is-expanded .p-icon--contextual-menu {
          transform: none;
        }
      }

      transform: rotate(180deg);
    }
  }

  &__name {
    display: inline-block;
    flex-grow: 1;
    padding: 0 1rem;
  }

  &__options {
    margin: 1rem 0;
    padding: 0;
  }

  &__link {
    display: block;
    padding: 0.25rem 1rem;

    &,
    &:visited {
      color: $color-light;
    }

    &:hover {
      background-color: $color-navigation-highlight;
    }
  }
}
