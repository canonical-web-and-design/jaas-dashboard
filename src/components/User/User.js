import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import classNames from "classnames";
import { useDispatch, useStore } from "react-redux";

import { logOut } from "app/actions";
import useSendAnalytics from "app/send-analytics-hook";

import "./_user.scss";

const User = () => {
  const sendAnalytics = useSendAnalytics();
  const [isExpanded, setIsExpanded] = useState(false);
  const dispatch = useDispatch();
  const getState = useStore().getState;

  useEffect(() => {
    if (isExpanded) {
      sendAnalytics({
        category: "User",
        action: "Opened user panel"
      });
    }
  }, [isExpanded, sendAnalytics]);

  return (
    <div
      className={classNames("user", {
        "is-expanded": isExpanded
      })}
    >
      <div className="user__header" onClick={() => setIsExpanded(!isExpanded)}>
        <i className="p-icon--user"></i>
        {/* @TODO Add user name here */}
        <span className="user__name">Account</span>
        <i className="p-icon--contextual-menu"></i>
      </div>
      <ul className="user__options">
        <li>
          <Link className="user__link" to="/settings">
            Settings
          </Link>
        </li>
        <li>
          <Link
            className="user__link"
            to="/"
            onClick={() => dispatch(logOut(getState))}
          >
            Log out
          </Link>
        </li>
      </ul>
    </div>
  );
};

export default User;
