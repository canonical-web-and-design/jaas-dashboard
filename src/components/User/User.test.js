import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { mount } from "enzyme";
import dataDump from "testing/complete-redux-store-dump";

import User from "./User";

const mockStore = configureStore([]);
describe("User Icon", () => {
  it("renders without crashing and matches snapshot", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <User />
        </Router>
      </Provider>
    );
    expect(wrapper.find(".user")).toMatchSnapshot();
  });

  it("toggles drop-down panel", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <Provider store={store}>
        <Router>
          <User />
        </Router>
      </Provider>
    );

    const user = ".user";
    const userOptionsToggle = ".user__header";

    expect(wrapper.find(user).hasClass("is-expanded")).toEqual(false);
    wrapper.find(userOptionsToggle).simulate("click");
    expect(wrapper.find(user).hasClass("is-expanded")).toEqual(true);
    wrapper.find(userOptionsToggle).simulate("click");
    expect(wrapper.find(user).hasClass("is-expanded")).toEqual(false);
  });
});
