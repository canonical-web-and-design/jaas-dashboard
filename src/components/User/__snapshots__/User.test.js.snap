// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Icon renders without crashing and matches snapshot 1`] = `
<div
  className="user"
>
  <div
    className="user__header"
    onClick={[Function]}
  >
    <i
      className="p-icon--user"
    />
    <span
      className="user__name"
    >
      Account
    </span>
    <i
      className="p-icon--contextual-menu"
    />
  </div>
  <ul
    className="user__options"
  >
    <li>
      <Link
        className="user__link"
        to="/settings"
      >
        <LinkAnchor
          className="user__link"
          href="/settings"
          navigate={[Function]}
        >
          <a
            className="user__link"
            href="/settings"
            onClick={[Function]}
          >
            Settings
          </a>
        </LinkAnchor>
      </Link>
    </li>
    <li>
      <Link
        className="user__link"
        onClick={[Function]}
        to="/"
      >
        <LinkAnchor
          className="user__link"
          href="/"
          navigate={[Function]}
          onClick={[Function]}
        >
          <a
            className="user__link"
            href="/"
            onClick={[Function]}
          >
            Log out
          </a>
        </LinkAnchor>
      </Link>
    </li>
  </ul>
</div>
`;
